<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
Great, Spencer! You've done a great job on all the comments and now your project has been accepted.
    
Thank you for your work and I wish you success in the following projects!

<div style="border:solid blue 2px; padding: 20px">
  
**Hello Spencer**

My name is Dima, and I will be reviewing your project. 

You will find my comments in coloured cells marked as 'Reviewer's comment'. The cell colour will vary based on the contents - I am explaining it further below. 

**Note:** Please do not remove or change my comments - they will help me in my future reviews and will make the process smoother for both of us. 

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment</b> 
    
Such comment will mark efficient solutions and good ideas that can be used in other projects.
</div>

<div class="alert alert-warning"; style="border-left: 7px solid gold">
<b>⚠️ Reviewer's comment</b> 
    
The parts marked with yellow comments indicate that there is room for optimisation. Though the correction is not necessary it is good if you implement it.
</div>

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment</b> 
    
If you see such a comment, it means that there is a problem that needs to be fixed. Please note that I won't be able to accept your project until the issue is resolved.
</div>

You are also very welcome to leave your comments / describe the corrections you've done / ask me questions, marking them with a different colour. You can use the example below: 

<div class="alert alert-info"; style="border-left: 7px solid blue">
<b>Student's comment</b>

## Basic Python - Project <a id='intro'></a>

<div style="border:solid green 2px; padding: 20px">
    
<div class="alert alert-success">
<b>Review summary</b> 
    
Spencer, thanks for submitting the project. You've done a very good job and I enjoyed reviewing it.
    
- You completed all the tasks.
- Your code was optimal and easy to read. 
- You wrote your own functions.
    
There are only a few critical comments that need to be corrected. You will find them in the red-colored cells in relevant sections. If you have any questions please write them when you return your project. 
    
I'll be looking forward to getting your updated notebook.

## Introduction <a id='intro'></a>
In this project, you will work with data from the entertainment industry. You will study a dataset with records on movies and shows. The research will focus on the "Golden Age" of television, which began in 1999 with the release of *The Sopranos* and is still ongoing.

The aim of this project is to investigate how the number of votes a title receives impacts its ratings. The assumption is that highly-rated shows (we will focus on TV shows, ignoring movies) released during the "Golden Age" of television also have the most votes.

### Stages 
Data on movies and shows is stored in the `/datasets/movies_and_shows.csv` file. There is no information about the quality of the data, so you will need to explore it before doing the analysis.

First, you'll evaluate the quality of the data and see whether its issues are significant. Then, during data preprocessing, you will try to account for the most critical problems.
 
Your project will consist of three stages:
 1. Data overview
 2. Data preprocessing
 3. Data analysis

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Title and introduction are essential parts of the project. Make sure you do not forget to include it in your further projects. 
    
It is optimal if introduction consists of:
    
- brief description of the situation;
- goal of the project;
- description of the data we are going to use.
</div>


## Stage 1. Data overview <a id='data_review'></a>

Open and explore the data.

You'll need `pandas`, so import it.

In [3]:
# importing pandas
import pandas as pd


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> Needed library has been added </div>

Read the `movies_and_shows.csv` file from the `datasets` folder and save it in the `df` variable:

In [4]:
# reading the files and storing them to df
df = pd.read_csv('/datasets/movies_and_shows.csv')


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> The correct path to the file is specified: the slash at the beginning of the path is very important, as it indicates that you need to search for the file in the root folder. </div>

Print the first 10 table rows:

In [5]:
# obtaining the first 10 rows from the df table
# hint: you can use head() and tail() in Jupyter Notebook without wrapping them into print()
df.head(10)

name                      Character   r0le        TITLE   Type  \
0   Robert De Niro                  Travis Bickle  ACTOR  Taxi Driver  MOVIE   
1     Jodie Foster                  Iris Steensma  ACTOR  Taxi Driver  MOVIE   
2    Albert Brooks                            Tom  ACTOR  Taxi Driver  MOVIE   
3    Harvey Keitel        Matthew 'Sport' Higgins  ACTOR  Taxi Driver  MOVIE   
4  Cybill Shepherd                          Betsy  ACTOR  Taxi Driver  MOVIE   
5      Peter Boyle                         Wizard  ACTOR  Taxi Driver  MOVIE   
6   Leonard Harris      Senator Charles Palantine  ACTOR  Taxi Driver  MOVIE   
7   Diahnne Abbott                Concession Girl  ACTOR  Taxi Driver  MOVIE   
8      Gino Ardito             Policeman at Rally  ACTOR  Taxi Driver  MOVIE   
9  Martin Scorsese  Passenger Watching Silhouette  ACTOR  Taxi Driver  MOVIE   

   release Year              genres  imdb sc0re  imdb v0tes  
0          1976  ['drama', 'crime']         8.2    808582.0  
1          1976  ['drama', 'crime']         8.2    808582.0  
2          1976  ['drama', 'crime']         8.2    808582.0  
3          1976  ['drama', 'crime']         8.2    808582.0  
4          1976  ['drama', 'crime']         8.2    808582.0  
5          1976  ['drama', 'crime']         8.2    808582.0  
6          1976  ['drama', 'crime']         8.2    808582.0  
7          1976  ['drama', 'crime']         8.2    808582.0  
8          1976  ['drama', 'crime']         8.2    808582.0  
9          1976  ['drama', 'crime']         8.2    808582.0

Obtain the general information about the table with one command:

In [6]:
# obtaining general information about the data in df
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 85579 entries, 0 to 85578
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0      name       85579 non-null  object 
 1   Character     85579 non-null  object 
 2   r0le          85579 non-null  object 
 3   TITLE         85578 non-null  object 
 4     Type        85579 non-null  object 
 5   release Year  85579 non-null  int64  
 6   genres        85579 non-null  object 
 7   imdb sc0re    80970 non-null  float64
 8   imdb v0tes    80853 non-null  float64
dtypes: float64(2), int64(1), object(6)
memory usage: 5.9+ MB


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Great - you've used a comprehensive set of methods to have a first look at the data.
    

The table contains nine columns. The majority store the same data type: object. The only exceptions are `'release Year'` (int64 type), `'imdb sc0re'` (float64 type) and `'imdb v0tes'` (float64 type). Scores and votes will be used in our analysis, so it's important to verify that they are present in the dataframe in the appropriate numeric format. Three columns (`'TITLE'`, `'imdb sc0re'` and `'imdb v0tes'`) have missing values.

According to the documentation:
- `'name'` — actor/director's name and last name
- `'Character'` — character played (for actors)
- `'r0le '` — the person's contribution to the title (it can be in the capacity of either actor or director)
- `'TITLE '` — title of the movie (show)
- `'  Type'` — show or movie
- `'release Year'` — year when movie (show) was released
- `'genres'` — list of genres under which the movie (show) falls
- `'imdb sc0re'` — score on IMDb
- `'imdb v0tes'` — votes on IMDb

We can see three issues with the column names:
1. Some names are uppercase, while others are lowercase.
2. There are names containing whitespace.
3. A few column names have digit '0' instead of letter 'o'. 


### Conclusions <a id='data_review_conclusions'></a> 

Each row in the table stores data about a movie or show. The columns can be divided into two categories: the first is about the roles held by different people who worked on the movie or show (role, name of the actor or director, and character if the row is about an actor); the second category is information about the movie or show itself (title, release year, genre, imdb figures).

It's clear that there is sufficient data to do the analysis and evaluate our assumption. However, to move forward, we need to preprocess the data.

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Please note that it is highly recommended to add a conclusion / summary after each section and describe briefly your observations and / or major outcomes of the analysis.

## Stage 2. Data preprocessing <a id='data_preprocessing'></a>
Correct the formatting in the column headers and deal with the missing values. Then, check whether there are duplicates in the data.

In [7]:
# the list of column names in the df table
df.columns

Index(['   name', 'Character', 'r0le', 'TITLE', '  Type', 'release Year',
       'genres', 'imdb sc0re', 'imdb v0tes'],
      dtype='object')

Change the column names according to the rules of good style:
* If the name has several words, use snake_case
* All characters must be lowercase
* Remove whitespace
* Replace zero with letter 'o'

In [8]:
# renaming columns
df = df.rename(columns={
    '   name': 'name',
    'Character': 'character',
    'r0le': 'role',
    'TITLE': 'title',
    '  Type': 'type',
    'release Year': 'release_year',
    'imdb sc0re': 'imbd_score',
    'imdb v0tes': 'imbd_votes',
})


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
This is a good way to rename the columns.

Check the result. Print the names of the columns once more:

In [9]:
# checking result: the list of column names
df.columns

Index(['name', 'character', 'role', 'title', 'type', 'release_year', 'genres',
       'imbd_score', 'imbd_votes'],
      dtype='object')

### Missing values <a id='missing_values'></a>
First, find the number of missing values in the table. To do so, combine two `pandas` methods:

In [10]:
# calculating missing values
print(df.isna().sum()) 


name               0
character          0
role               0
title              1
type               0
release_year       0
genres             0
imbd_score      4609
imbd_votes      4726
dtype: int64


<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
The isna() method is selected to find the missing values, it's great!

Not all missing values affect the research: the single missing value in `'title'` is not critical. The missing values in columns `'imdb_score'` and `'imdb_votes'` represent around 6% of all records (4,609 and 4,726, respectively, of the total 85,579). This could potentially affect our research. To avoid this issue, we will drop rows with missing values in the `'imdb_score'` and `'imdb_votes'` columns.

In [11]:
# dropping rows where columns with title, scores and votes have missing values
df['title'] = df['title'].fillna('unknown')
df = df.dropna(subset = ['imbd_score'])
df = df.dropna(subset = ['imbd_votes'])

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Perfect!

Make sure the table doesn't contain any more missing values. Count the missing values again.

In [12]:
# counting missing values
df.isna().sum()


name            0
character       0
role            0
title           0
type            0
release_year    0
genres          0
imbd_score      0
imbd_votes      0
dtype: int64

### Duplicates <a id='duplicates'></a>
Find the number of duplicate rows in the table using one command:

In [13]:
# counting duplicate rows
df.duplicated().sum()

6994

Review the duplicate rows to determine if removing them would distort our dataset.

In [14]:
# Produce table with duplicates (with original rows included) and review last 5 rows
duplicate_rows = df[df.duplicated()]
duplicate_rows.tail()

name                 character   role    title   type  \
85569       Jessica Cediel           Liliana Navarro  ACTOR  Lokillo  MOVIE   
85570  Javier Gardeaz?­bal  Agust??n "Peluca" Ort??z  ACTOR  Lokillo  MOVIE   
85571        Carla Giraldo            Valery Reinoso  ACTOR  Lokillo  MOVIE   
85572  Ana Mar??a S?­nchez                   Lourdes  ACTOR  Lokillo  MOVIE   
85577         Isabel Gaona                    Cacica  ACTOR  Lokillo  MOVIE   

       release_year      genres  imbd_score  imbd_votes  
85569          2021  ['comedy']         3.8        68.0  
85570          2021  ['comedy']         3.8        68.0  
85571          2021  ['comedy']         3.8        68.0  
85572          2021  ['comedy']         3.8        68.0  
85577          2021  ['comedy']         3.8        68.0

There are two clear duplicates in the printed rows. We can safely remove them.
Call the `pandas` method for getting rid of duplicate rows:

In [15]:
# removing duplicate rows
df.drop_duplicates(inplace=True)

Check for duplicate rows once more to make sure you have removed all of them:

In [16]:
# checking for duplicates
df.duplicated().sum()

0

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Great, you found and removed the duplicates. And did very thorough checks to make sure the duplicates are gone.

Now get rid of implicit duplicates in the `'type'` column. For example, the string `'SHOW'` can be written in different ways. These kinds of errors will also affect the result.

Print a list of unique `'type'` names, sorted in alphabetical order. To do so:
* Retrieve the intended dataframe column 
* Apply a sorting method to it
* For the sorted column, call the method that will return all unique column values

In [20]:
# viewing unique type names
unsorted_types = df['type']
sorted_types = unsorted_types.sort_values()
unique_types = sorted_types.unique()
sorted__unique_types = sorted(unique_types)
sorted__unique_types

['MOVIE',
 'SHOW',
 'movies',
 'shows',
 'the movie',
 'tv',
 'tv series',
 'tv show',
 'tv shows']

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 

Please note, that according to the technical task it was asked:
    
* For the **sorted** column, call the method that will return all unique column values
    
So, we should use sorting here

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
Perfect!

Look through the list to find implicit duplicates of `'show'` (`'movie'` duplicates will be ignored since the assumption is about shows). These could be names written incorrectly or alternative names of the same genre.

You will see the following implicit duplicates:
* `'shows'`
* `'SHOW'`
* `'tv show'`
* `'tv shows'`
* `'tv series'`
* `'tv'`

To get rid of them, declare the function `replace_wrong_show()` with two parameters: 
* `wrong_shows_list=` — the list of duplicates
* `correct_show=` — the string with the correct value

The function should correct the names in the `'type'` column from the `df` table (i.e., replace each value from the `wrong_shows_list` list with the value in `correct_show`).

In [21]:
# function for replacing implicit duplicates
def replace_wrong_show(df, wrong_shows_list, correct_show):
    # Replace incorrect values in 'type' column with correct value
    df['type'] = df['type'].replace(wrong_shows_list, correct_show)

Call `replace_wrong_show()` and pass it arguments so that it clears implicit duplicates and replaces them with `SHOW`:

In [22]:
# removing implicit duplicates
wrong_shows_list = ['shows', 'SHOW', 'tv show', 'tv shows', 'tv series', 'tv']
correct_show = 'SHOW'

df['type'] = df['type'].replace(wrong_shows_list, correct_show)

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Yes, this is what was needed!

Make sure the duplicate names are removed. Print the list of unique values from the `'type'` column:

In [23]:
# viewing unique genre names
df['type'].unique()

array(['MOVIE', 'the movie', 'SHOW', 'movies'], dtype=object)

### Conclusions <a id='data_preprocessing_conclusions'></a>
We detected three issues with the data:

- Incorrect header styles
- Missing values
- Duplicate rows and implicit duplicates

The headers have been cleaned up to make processing the table simpler.

All rows with missing values have been removed. 

The absence of duplicates will make the results more precise and easier to understand.

Now we can move on to our analysis of the prepared data.

## Stage 3. Data analysis <a id='hypotheses'></a>

Based on the previous project stages, you can now define how the assumption will be checked. Calculate the average amount of votes for each score (this data is available in the `imdb_score` and `imdb_votes` columns), and then check how these averages relate to each other. If the averages for shows with the highest scores are bigger than those for shows with lower scores, the assumption appears to be true.

Based on this, complete the following steps:

- Filter the dataframe to only include shows released in 1999 or later.
- Group scores into buckets by rounding the values of the appropriate column (a set of 1-10 integers will help us make the outcome of our calculations more evident without damaging the quality of our research).
- Identify outliers among scores based on their number of votes, and exclude scores with few votes.
- Calculate the average votes for each score and check whether the assumption matches the results.

To filter the dataframe and only include shows released in 1999 or later, you will take two steps. First, keep only titles published in 1999 or later in our dataframe. Then, filter the table to only contain shows (movies will be removed).

In [24]:
# using conditional indexing modify df so it has only titles released after 1999 (with 1999 included)

# give the slice of dataframe new name
later_titles = df[df['release_year'] >= 1999]
later_titles

name            character      role     title       type  \
1664          Jeff Probst       Himself - Host     ACTOR  Survivor       SHOW   
1955     Benicio del Toro  Franky Four Fingers     ACTOR    Snatch      MOVIE   
1956        Dennis Farina           Cousin Avi     ACTOR    Snatch      MOVIE   
1957         Vinnie Jones    Bullet Tooth Tony     ACTOR    Snatch      MOVIE   
1958            Brad Pitt        Mickey O'Neil     ACTOR    Snatch      MOVIE   
...                   ...                  ...       ...       ...        ...   
85573       A??da Morales              Maritza     ACTOR   Lokillo  the movie   
85574    Adelaida Buscato           Mar??a Paz     ACTOR   Lokillo  the movie   
85575  Luz Stella Luengas         Karen Bayona     ACTOR   Lokillo  the movie   
85576        In??s Prieto                Fanny     ACTOR   Lokillo  the movie   
85578      Julian Gaviria              unknown  DIRECTOR   Lokillo  the movie   

       release_year               genres  imbd_score  imbd_votes  
1664           2000          ['reality']         7.4     24687.0  
1955           2000  ['crime', 'comedy']         8.3    841435.0  
1956           2000  ['crime', 'comedy']         8.3    841435.0  
1957           2000  ['crime', 'comedy']         8.3    841435.0  
1958           2000  ['crime', 'comedy']         8.3    841435.0  
...             ...                  ...         ...         ...  
85573          2021           ['comedy']         3.8        68.0  
85574          2021           ['comedy']         3.8        68.0  
85575          2021           ['comedy']         3.8        68.0  
85576          2021           ['comedy']         3.8        68.0  
85578          2021           ['comedy']         3.8        68.0  

[69881 rows x 9 columns]

In [26]:
# repeat conditional indexing so df has only shows (movies are removed as result)
later_titles = later_titles[later_titles['type'] == 'SHOW']


<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 
    
Not quite right. Please note, that we should filter `SHOW` in the new dataframe **later_titles**

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
Yep, well done!

The scores that are to be grouped should be rounded. For instance, titles with scores like 7.8, 8.1, and 8.3 will all be placed in the same bucket with a score of 8.

In [27]:
# rounding column with scores
later_titles = later_titles.round()

#checking the outcome with tail()
later_titles.tail()

name character      role      title  type  release_year  \
85433     Maneerat Kam-Uan        Ae     ACTOR  Let's Eat  SHOW          2021   
85434    Rudklao Amratisha   unknown     ACTOR  Let's Eat  SHOW          2021   
85435      Jaturong Mokjok   unknown     ACTOR  Let's Eat  SHOW          2021   
85436     Pisamai Wilaisak   unknown     ACTOR  Let's Eat  SHOW          2021   
85437  Sarawut Wichiensarn   unknown  DIRECTOR  Let's Eat  SHOW          2021   

                    genres  imbd_score  imbd_votes  
85433  ['drama', 'comedy']         8.0         5.0  
85434  ['drama', 'comedy']         8.0         5.0  
85435  ['drama', 'comedy']         8.0         5.0  
85436  ['drama', 'comedy']         8.0         5.0  
85437  ['drama', 'comedy']         8.0         5.0

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Great!

It is now time to identify outliers based on the number of votes.

In [34]:
# Use groupby() for scores and count all unique values in each group, print the result
grouping_score = later_titles.groupby('imbd_score')['imbd_votes'].count()
grouping_score

imbd_score
2.0       24
3.0       27
4.0      180
5.0      592
6.0     2494
7.0     4706
8.0     4842
9.0      557
10.0       8
Name: imbd_votes, dtype: int64

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 

It's better not get 8 columns of repeating values. Just groupby on two columns together

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
It was performed absolutely correctly

Based on the aggregation performed, it is evident that scores 2 (24 voted shows), 3 (27 voted shows), and 10 (only 8 voted shows) are outliers. There isn't enough data for these scores for the average number of votes to be meaningful.

To obtain the mean numbers of votes for the selected scores (we identified a range of 4-9 as acceptable), use conditional filtering and grouping.

In [35]:
# filter dataframe using two conditions (scores to be in the range 4-9)
later_titles = later_titles[(later_titles['imbd_score'] >= 4) & (later_titles['imbd_score'] <= 9)]
# group scores and corresponding average number of votes, reset index and print the result
score_vote = later_titles.groupby('imbd_score')['imbd_votes'].mean().reset_index()
score_vote

imbd_score     imbd_votes
0         4.0    5277.583333
1         5.0    3143.942568
2         6.0    3481.717322
3         7.0    8727.068211
4         8.0   30299.460967
5         9.0  126904.109515

<div class="alert alert-danger"; style="border-left: 7px solid red">
<b>⛔️ Reviewer's comment, v. 1</b> 

And here also, please filter `later_titles`, not `df`

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 2</b> 
    
Awesome!

Now for the final step! Round the column with the averages, rename both columns, and print the dataframe in descending order.

In [70]:
# round column with averages
score_vote = score_vote.round()
# rename columns
score_vote = score_vote.rename(columns={
    'imbd_score': 'IMBD Score',
    'imbd_votes': 'IMBD Votes'
})
# print dataframe in descending order
Votes = score_vote.sort_values(by='IMBD Votes', ascending = False)
Votes

IMBD Score  IMBD Votes
5         9.0    126904.0
4         8.0     30299.0
3         7.0      8727.0
0         4.0      5278.0
2         6.0      3482.0
1         5.0      3144.0

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Great! Also correct rounding and grouping in this section

The assumption macthes the analysis: the shows with the top 3 scores have the most amounts of votes.

## Conclusion <a id='hypotheses'></a>

The research done confirms that highly-rated shows released during the "Golden Age" of television also have the most votes. While shows with score 4 have more votes than ones with scores 5 and 6, the top three (scores 7-9) have the largest number. The data studied represents around 94% of the original set, so we can be confident in our findings.

<div class="alert alert-success"; style="border-left: 7px solid green">
<b>✅ Reviewer's comment, v. 1</b> 
    
Overall conclusion is an important part, where we should include the summary of the outcomes of the project.